promotion/datastore.go

```diff
@@ -1010,6 +1056,154 @@ limit 1`
+	MintDrainJobPending = "pending"
+	// MintDrainJobFailed - failed status for the mint_drain job
+	MintDrainJobFailed = "failed"
+	// MintDrainJobComplete - complete status for the mint_drain job
+	MintDrainJobComplete = "complete"
+)
+
+// RunNextMintDrainJob to process mints vg
+func (pg *Postgres) RunNextMintDrainJob(ctx context.Context, worker MintWorker) (bool, error) {
+
+	// setup a logger
+	logger, err := appctx.GetLogger(ctx)
+	if err != nil {
+		// no logger, setup
+		ctx, logger = logging.SetupLogger(ctx)
+	}
+
+	// get and parse the correct transfer promotion id to create claims on
+	braveTransferPromotionIDs, ok := ctx.Value(appctx.BraveTransferPromotionIDCTXKey).([]string)
+	if !ok {
+		logger.Error().Err(errMissingTransferPromotion).
+			Msg("MintJob: missing transfer promotion id")
+		return false, errMissingTransferPromotion
+	}
+
+	tx, err := pg.RawDB().Beginx()
+	attempted := false
+	if err != nil {
+		return attempted, err
+	}
+	defer pg.RollbackTx(tx)
+
+	// get the mint job.
+	statement := `
+select *
+from mint_drain
+where not erred and status = 'pending'
+for update skip locked
+limit 1`
+
+	job := MintDrainJob{}
+	err = tx.Get(&job, statement)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return attempted, nil
+		}
+		return attempted, err
+	}
+	// are all of the claims associated with all of the promotions drained?
+
+	statement = `
+select
+	true
+from
+	mint_drain_promotion mdp
+	join claims c
+		on (c.promotion_id=mdp.promotion_id)
+where
+	mdp.mint_drain_id = $1
+	and c.drained = false
+	and c.wallet_id= $2
+limit 1`
+	var drained bool
+	err = tx.Get(&drained, statement, job.ID, job.WalletID)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return attempted, nil
+		}
+		return attempted, err
+	}
+
+	if !drained {
+		return attempted, nil
+	}
+
+	// yes? set status to complete and mint the grants
+	promoIDs, err := toUUIDs(braveTransferPromotionIDs...)
+	if err != nil {
+		// log the error from redeem and transfer
+		logger.Error().Err(err).Msg("failed to derive promotion ids from configuration")
+		return attempted, err
+	}
+
+	attempted = true
+	// mint the grant to the wallet's deposit destination
+	statement = `
+select
+	user_deposit_destination
+from
+	wallets
+where
+	id = $1
+`
+	var depositDestination string
+	err = tx.Get(&depositDestination, statement, job.WalletID)
+	if err != nil {
+		return attempted, err
+	}
+
+	if depositDestination == "" {
+		return attempted, errors.New("Wallet is not verified")
+	}
+
+	depositDestinationUUID, err := uuid.FromString(depositDestination)
+	if err != nil {
+		return attempted, errors.New("destination invalid wallet id")
+	}
+
+	err = worker.MintGrant(ctx, depositDestinationUUID, job.Total, promoIDs...)
+	if err != nil {
+		// log the error from redeem and transfer
+		logger.Error().Err(err).Msg("failed to mint grants")
+		if _, err := tx.Exec(`update mint_drain set erred = true where id = $1`, job.ID); err != nil {
```

Comment:
this looks a little funny. are we trying to only send the mintgrant error back? and why would we commit on the next line?

Reply:
Yes, we want to commit the transaction to update the mint_drain to set the errored status prior to returning.